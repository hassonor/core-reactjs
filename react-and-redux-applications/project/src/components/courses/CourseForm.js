import React from "react";
import PropTypes from "prop-types";
import TextInput from "../common/TextInput";
import SelectInput from "../common/SelectInput";

const CourseForm = ({
                        course,
                        authors,
                        onSave,
                        onChange,
                        saving = false,
                        errors = {}
                    }) => {
    return (
        <form onSubmit={onSave}>
            <h2>{course.id ? "Edit" : "Add"} Course</h2>
            {errors.onSave && (
                <div className="alert alert-danger" role="alert">
                    {errors.onSave}
                </div>
            )}
            <TextInput
                name="title"
                label="Title"
                value={course.title}
                onChange={onChange}
                error={errors.title}
            />

            <SelectInput
                name="authorId"
                label="Author"
                value={course.authorId || ""}
                defaultOption="Select Author"
                options={authors.map(author => ({
                    value: author.id,
                    text: author.name
                }))}
                onChange={onChange}
                error={errors.author}
            />

            <TextInput
                name="category"
                label="Category"
                value={course.category}
                onChange={onChange}
                error={errors.category}
            />

            <button type="submit" disabled={saving} className="btn btn-primary">
                {saving ? "Saving..." : "Save"}
            </button>
        </form>
    );
};

CourseForm.propTypes = {
    authors: PropTypes.array.isRequired,
    course: PropTypes.object.isRequired,
    errors: PropTypes.object,
    onSave: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    saving: PropTypes.bool
};

export default CourseForm;
