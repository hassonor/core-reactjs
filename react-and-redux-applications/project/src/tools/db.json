{
  "courses": [
    {
      "id": 1,
      "title": "Securing React Apps with Auth03",
      "slug": "react-auth0-authentication-security",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 2,
      "title": "React: The Big Picture",
      "slug": "react-big-picture",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 3,
      "title": "Creating Reusable React Components",
      "slug": "react-creating-reusable-components",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 5,
      "title": "Building Applications with React and Redux",
      "slug": "react-redux-react-router-es6",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 7,
      "title": "Clean Code: Writing Code for Humans",
      "slug": "writing-clean-code-humans",
      "authorId": 1,
      "category": "Software Practices"
    },
    {
      "id": 8,
      "title": "Architecting Applications for the Real World",
      "slug": "architecting-applications-dotnet",
      "authorId": 1,
      "category": "Software Architecture"
    },
    {
      "id": 9,
      "title": "Becoming an Outlier: Reprogramming the Developer Mind",
      "slug": "career-reboot-for-developer-mind",
      "authorId": 1,
      "category": "Career"
    },
    {
      "id": 10,
      "title": "Web Component Fundamentals",
      "slug": "web-components-shadow-dom",
      "authorId": 1,
      "category": "HTML5"
    },
    {
      "title": "Or Hasson",
      "authorId": 2,
      "category": "1222",
      "createdAt": 1647168869835,
      "slug": "or-hasson",
      "id": 11
    }
  ],
  "authors": [
    {
      "id": 1,
      "name": "Cory House"
    },
    {
      "id": 2,
      "name": "Or Hasson"
    },
    {
      "id": 3,
      "name": "Avi Nimni"
    }
  ]
}