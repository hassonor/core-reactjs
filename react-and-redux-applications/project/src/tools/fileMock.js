// Mocks file imports for Jest. As suggested by https://jestjs.io/docs/en/webpack
module.exports = "test-file-stub";
