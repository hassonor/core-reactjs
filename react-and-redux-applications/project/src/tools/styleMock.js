// Mocks CSS imports for Jest. As suggested by https://jestjs.io/docs/en/webpack
module.exports = {};
