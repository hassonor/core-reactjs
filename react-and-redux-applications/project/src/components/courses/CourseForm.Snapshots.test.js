import React from "react";
import CourseForm from "./CourseForm";
import renderer from "react-test-renderer";
import {courses, authors} from "../../tools/mockData";

it("sets submit button label 'Saving...' when saving is true", () => {
    const tree = renderer.create(
        <CourseForm
            course={courses[0]}
            authors={authors}
            onSave={jest.fn()}
            onChange={jest.fn()}
            saving
        />);

    expect(tree).toMatchSnapshot();
})


it("sets submit button label 'Save' when saving is false", () => {
    const tree = renderer.create(
        <CourseForm
            course={courses[0]}
            authors={authors}
            onSave={jest.fn()}
            onChange={jest.fn()}
            saving={false}
        />);

    expect(tree).toMatchSnapshot();
})