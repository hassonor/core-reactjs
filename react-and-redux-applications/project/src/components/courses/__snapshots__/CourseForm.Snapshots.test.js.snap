// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets submit button label 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Course
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="title"
        onChange={[MockFunction]}
        type="text"
        value="Securing React Apps with Auth0"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="authorId"
    >
      Author
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="authorId"
        onChange={[MockFunction]}
        value={1}
      >
        <option
          value=""
        >
          Select Author
        </option>
        <option
          value={1}
        >
          Cory House
        </option>
        <option
          value={2}
        >
          Or Hasson
        </option>
        <option
          value={3}
        >
          Avi Nimni
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="category"
    >
      Category
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="category"
        onChange={[MockFunction]}
        type="text"
        value="JavaScript"
      />
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets submit button label 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Course
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="title"
        onChange={[MockFunction]}
        type="text"
        value="Securing React Apps with Auth0"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="authorId"
    >
      Author
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="authorId"
        onChange={[MockFunction]}
        value={1}
      >
        <option
          value=""
        >
          Select Author
        </option>
        <option
          value={1}
        >
          Cory House
        </option>
        <option
          value={2}
        >
          Or Hasson
        </option>
        <option
          value={3}
        >
          Avi Nimni
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="category"
    >
      Category
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="category"
        onChange={[MockFunction]}
        type="text"
        value="JavaScript"
      />
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    type="submit"
  >
    Saving...
  </button>
</form>
`;
